lib/src/main/java/graphql/nadel/validation/NadelHydrationArgumentValidation.kt

```diff
@@ -0,0 +1,288 @@
+            val unwrappedActorFieldArgType = actorFieldArgType.unwrapNonNull()
+
+            //could have ID feed into [ID] (as a possible batch hydration case).
+            // in this case we need to unwrap the list and check types
+            if (isBatchHydration && (!unwrappedHydrationSourceFieldType.isList && unwrappedActorFieldArgType.isList)) {
+                val error = getHydrationInputErrors(
+                        unwrappedHydrationSourceFieldType,
+                        unwrappedActorFieldArgType.unwrapOne(),
+                        parent,
+                        overallField,
+                        remoteArg,
+                        hydration,
+                        actorFieldName
+                )
+                if (error != null) {
+                    return NadelSchemaValidationError.IncompatibleHydrationArgumentType(
+                            parent,
+                            overallField,
+                            remoteArg,
+                            hydrationSourceFieldType,
+                            actorFieldArgType,
+                            actorFieldName
+                    )
+                }
+            }
+            //could have [ID] feed into ID (non-batch, ManyToOne case explained in NadelHydrationStrategy.kt)
+            // in this case we need to unwrap the list and check types
+            else if (!isBatchHydration && (unwrappedHydrationSourceFieldType.isList && !unwrappedActorFieldArgType.isList)) {
+                val error = getHydrationInputErrors(
+                        unwrappedHydrationSourceFieldType.unwrapOne(),
+                        unwrappedActorFieldArgType,
+                        parent,
+                        overallField,
+                        remoteArg,
+                        hydration,
+                        actorFieldName
+                )
+                if (error != null) {
+                    return NadelSchemaValidationError.IncompatibleHydrationArgumentType(
+                            parent,
+                            overallField,
+                            remoteArg,
+                            hydrationSourceFieldType,
+                            actorFieldArgType,
+                            actorFieldName
+                    )
+                }
+
+            }
+            // Otherwise we can just check the types normally
+            else {
+                return getHydrationInputErrors(
+                        hydrationSourceFieldType,
+                        actorFieldArgType,
+                        parent,
+                        overallField,
+                        remoteArg,
+                        hydration,
+                        actorFieldName
+                )
+            }
+            return null
+        }
+
+        private fun getHydrationInputErrors(
+                hydrationSourceFieldType: GraphQLType,
```

Comment:
This isn't strictly correct right? This can be an argument's type right?

I think we should rename this into `sourceType` vs `actorType` (throughout the file).

Reply:
Yeah, I am for renaming to `hydrationSourceType`, as it is not always from a field.